// @bootstrap-styled/v4/lib/Dt/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.propTypes = exports.defaultProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _gridFramework = require("@bootstrap-styled/css-mixins/lib/grid-framework");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Description Term
 *
 */
var defaultProps = {
  theme: {
    '$dt-font-weight': 'bold',
    '$grid-columns': '12',
    '$grid-gutter-width': '30px',
    '$grid-breakpoints': {
      xs: '0',
      sm: '576px',
      md: '768px',
      lg: '992px',
      xl: '1200px'
    },
    '$enable-grid-classes': true
  }
};
exports.defaultProps = defaultProps;
var propTypes = {
  /** Theme variables. Can be: */
  theme: _propTypes.default.shape({
    '$dt-font-weight': _propTypes.default.string,
    '$grid-columns': _propTypes.default.string,
    '$grid-gutter-width': _propTypes.default.string,
    '$grid-breakpoints': _propTypes.default.object,
    '$enable-grid-classes': _propTypes.default.bool
  })
};
/**
 * Use our `<Dt />` component defines a term/name in a description list.
 */

exports.propTypes = propTypes;

var Dt = _styledComponents.default.dt.withConfig({
  displayName: "Dt",
  componentId: "sc-5rnls5-0"
})(["  ", ""], function (props) {
  return "\n    /* Reboot Scss */\n    font-weight: ".concat(props.theme['$dt-font-weight'], ";\n    ").concat((0, _gridFramework.makeGridColumns)(props.theme['$enable-grid-classes'], props.theme['$grid-columns'], props.theme['$grid-gutter-width'], props.theme['$grid-breakpoints']), "\n  ");
});

Dt.defaultProps = defaultProps;
Dt.propTypes = propTypes;
/** @component */

var _default = Dt;
exports.default = _default;
